python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,154 @@
+# under the License.
+"""FileIO implementation for reading and writing table files that uses pyarrow.fs
+
+This file contains a FileIO implementation that relies on the filesystem interface provided
+by PyArrow. It relies on PyArrow's `from_uri` method that infers the correct filesystem
+type to use. Theoretically, this allows the supported storage types to grow naturally
+with the pyarrow library.
+"""
+
+import os
+from typing import Union
+from urllib.parse import urlparse
+
+from pyarrow.fs import FileSystem, FileType
+
+from iceberg.io.base import FileIO, InputFile, InputStream, OutputFile, OutputStream
+
+
+class PyArrowFile(InputFile, OutputFile):
+    """A combined InputFile and OutputFile implementation that uses a pyarrow filesystem to generate pyarrow.lib.NativeFile instances
+
+    Args:
+        location(str): A URI or a path to a local file
+
+    Attributes:
+        location(str): The URI or path to a local file for a PyArrowFile instance
+        exists(bool): Whether the file exists or not
+        filesystem(pyarrow.fs.FileSystem): An implementation of the FileSystem base class inferred from the location
+
+    Examples:
+        >>> from iceberg.io.pyarrow import PyArrowFile
+        >>> input_file = PyArrowFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()  # Read the contents of the PyArrowFile instance
+        >>> output_file = PyArrowFile("s3://baz/qux.txt")
+        >>> output_file.create().write(b'foobytes')  # Write bytes to the PyArrowFile instance
+    """
+
+    def __init__(self, location: str):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+        if not parsed_location.scheme:  # If no scheme, assume the path is to a local file
+            self._filesystem, self._path = FileSystem.from_uri(os.path.abspath(location))
+        else:
+            self._filesystem, self._path = FileSystem.from_uri(location)  # Infer the proper filesystem
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file = self._filesystem.open_input_file(self._path)
+        return file.size()
+
+    @property
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
+        file_info = self._filesystem.get_file_info(self._path)
+        return False if file_info.type == FileType.NotFound else True
+
+    def open(self) -> InputStream:
+        """Opens the location using a PyArrow FileSystem inferred from the location
+
+        Returns:
+            pyarrow.lib.NativeFile: A NativeFile instance for the file located at `self.location`
+        """
+        input_file = self._filesystem.open_input_file(self._path)
+        if not isinstance(input_file, InputStream):
+            raise TypeError(
+                f"Object of type {type(input_file)} returned from PyArrowFile.open does not match the InputStream protocol."
+            )
```

Comment:
When do we expect this to happen? [`open_input_file`](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.open_input_file) states that it returns `NativeFile`, which we know is an instance of the `InputStream` protocol. I think this check probably makes sense in tests, but I would not ever expect a failure here at runtime.

Reply:
You're right I think it's good enough to just include this in tests. I was thinking we would catch a change in the behavior of what pyarrow returns from `open_input_file`, but that would be something we catch in tests after updating the version or something along those lines, not at runtime.

Reply:
Fixed in [this](https://github.com/apache/iceberg/pull/4081/commits/7fa28c766c3dd5bd6702abcf4d85860c6bbb164c) commit.